### PR TITLE
fix formatting for typescript Enum with computed members

### LINKED ISSUE #12926: Syntax error after formatting when using computed key in TypeScript enum
**Prettier 2.6.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuECCuBbABAQS8AHSiywG0BLAakoF0iBfEAGhAgAcZzoBnZUAQwBOgiAHcACkIS8U-ADaj+AT14sARoP5gA1nBgBlfhjgAZclDjIAZvO5x1mnXv1st5gObIYgtPdQY1OAATIOCTfih3NH53OAAxCEEMfhhOSOQQfjQYCGYQAAsYDDkAdXzyeG5XMDh9aQryADcKpQywblUQcztBGHFNd2TrWz8AK24AD30POTgARTQIeGG5OxZXQR6MmCU2OG4wQXIOPLYj2BLyIJh85AAOAAZ1kTsSzTYMs-24QUbLFgAjot4P12DJMtwALQWYLBPKCOBA8gI-oxIZIGyrPx2DDkLw+bEzebAywYkYsGD8NSXa63JAAJgpmnIcg8AGEIBh0ahuABWPJoOwAFSpMkxaxAjV8AEkoKFYPpDscYDg5fodrMVnZ6PQgA)

```sh
--parser typescript
```

**Input 1:**

```tsx
enum A {
  [i++]
}
```

**Output 1:**

```tsx
enum A {
  i++,
}

```
**Input 2:**

```tsx
const a = "b"
enum A {
  [a] = 2
}
```

**Output 2:**

```tsx
const a = "b";
enum A {
  a = 2,
}

```

**Expected behavior:**
The output should be valid syntax. Although TypeScript will error during compiling that computed keys are not allowed in enums, they are valid syntax and parse correctly. Prettier should still support formatting this. Note that TypeScript still produces emit even when erroring, meaning the code would be able to run.

**Actual behavior:**
The output incorrectly removes the brackets, producing a syntax error on example 1, or modifying the runtime behaviour in example 2.

### PR DESCRIPTION
**Pull request by HosokawaR**
## Description

Fixes #12926  

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: changelog_unreleased/typescript/12930.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+const bar = "bar"
+enum B {
+  [bar] = 2,
+}
+
+const foo = () => "foo";
+enum C {
+  [foo()] = 2,
+}
```
Comment (sosukesuzuki) [inline changelog_unreleased/typescript/12930.md:18]: I think just the first one example should suffice.

Comment (HosokawaR) [reply changelog_unreleased/typescript/12930.md]: Surely. I removed extra examples. 
Thank you !

**Review by fisker — APPROVED** [review]

**Review by sosukesuzuki — APPROVED** [review]

**Comment by Dionysusnu** [comment]
This means
```ts
enum A {
  ["a"] = 2
}
```
no longer gets turned into
```ts
enum A {
  "a" = 2
}
```
like it did before (a string literal was the only case to get valid output)
Has any consideration been given for in general turning these into:
```ts
enum A {
  a = 2
}
```
In those cases where `a` is a valid identifier, or
```ts
enum A {
  "invalidIdentifierInNormalRules" = 2
}
```
for invalid identifiers?

**Comment by HosokawaR** [comment]
@Dionysusnu 

Sorry for the late reply.

> Has any consideration been given for in general turning these into

No, I haven't. In my opinion, these transformations go beyond the role of a prettier.

Actually, `enum A { ["a"] = 2 }`, `enum A { "a" = 2 }` and `enum A { a = 2 }` are converted into same JavaScript. (Example [here](https://www.typescriptlang.org/play?#code/KYOwrgtgBAglDeUDaAiAhigulAvFATFAL4DcAUGaJFAEIJToq4HHmXjQDC9azhpZIA).)

However, I couldn't find the documents which say these `enum` have the same meaning and are interchangeable with each other.

So, I thought these transformations go beyond the role of prettier as a formatter.

**Pull request merged**